### PR TITLE
JP-4178: Follow-up on test failures

### PR DESCRIPTION
**Pull request by tapastro**
Partially addresses [JP-4178](https://jira.stsci.edu/browse/JP-4178)

This PR attempts to address test failures seen after merging https://github.com/spacetelescope/jwst/pull/10022, see: https://github.com/spacetelescope/RegressionTests/actions/runs/21733331044

It appears as though handling of tracemodels was altered - this might be the minimal required fix, but it's not clear that behavior hasn't fundamentally changed, i.e. that tracemodels are not being stored in the way they were prior to the PR merge. More investigation may be helpful.

## Tasks
- [x] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
    - if your change breaks **step-level or public API** ([as defined in the docs](https://jwst.readthedocs.io/en/latest/jwst/user_documentation/more_information.html#api-public-vs-private)), also add a `changes/<PR#>.breaking.rst` news fragment
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by tapastro** [comment]
RTs : https://github.com/spacetelescope/RegressionTests/actions/runs/21739446988

File: jwst/extract_1d/soss_extract/soss_extract.py
Comment (tapastro) [inline jwst/extract_1d/soss_extract/soss_extract.py:1763]: I think this is the root cause - prior to Tim's PR, all_tracemodels would be an empty dict, and no assignment would be attempted. My hasty fix was to check for valid arrays on prior to assignment below, but alternatively we could see if returning all_tracemodels behavior back to an emtpy dict would be preferable. The tests that returned errors are now passing.

**Comment by melanieclarke** [comment]
Wondering about the lessons learned, here... How did we miss this before merge?  Did we not run regtests on that branch?

**Comment by melanieclarke** [comment]
There is a slow unit test failure in the regtests that still needs fixing: test_extract_niriss_soss_96, in jwst.extract_1d.tests.test_extract_1d_step.

**Comment by tapastro** [comment]
The unit test is not failing locally. I'll see if passing zero values is sufficient.

These got through because of a communication failure - I thought the PR was ready to go apart from INS approval and my rubber stamp. I needed to re-run RTs prior to that stamp.

File: jwst/extract_1d/soss_extract/soss_extract.py
Comment (emolter) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: I don't think your first statement is true.  Prior to Tim's PR the line `    all_tracemodels = {order: [tracemodels[order]] for order in tracemodels}` appears after running the first integration but before running all the rest.

I'm running the regtests locally to see if I agree that this is an appropriate fix.

**Comment by emolter** [comment]
For my part, I don't think I ever took a step back to review the PR "as a whole" after making my rounds of comments, I only checked to see if those comments were being resolved.

File: jwst/extract_1d/soss_extract/soss_extract.py
Comment (emolter) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: The tracemodels are only saved if extra optional outputs are requested, and I confirmed that they look to be produced as expected in that case.  This error is happening when optional outputs were _not_ requested.  I think an alternative here would be to wrap the `for order in all_tracemodels` loop with an `if generate_model:`.  That would be more descriptive of why the attribute setting is skipped in some cases.

Comment (emolter) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: I'm wondering though if there could be some weird case where the tracemodel would still not be 3-D and fail validation against `ndim`, even if `generate_model` is True.  So your proposed fix might be safer, but then again it might be more likely to hide a different bug.

File: jwst/extract_1d/tests/test_extract_1d_step.py [hunk 1]
```diff
@@ -205,7 +205,7 @@ def test_extract_niriss_soss_96(tmp_path, mock_niriss_soss_96):
 
     # output flux and errors are non-zero, exact values will depend
     # on extraction parameters
-    assert np.all(result.spec[0].spec_table["FLUX"] > 0)
+    assert np.all(result.spec[0].spec_table["FLUX"] >= 0)
```
Comment (melanieclarke) [inline jwst/extract_1d/tests/test_extract_1d_step.py]: This test would pass if all the flux values were zero, which is not quite the intent of the test.  Checking locally on main, all values are around 40, so it's confusing that this would be failing in RT.  It shouldn't be a precision issue, unless there's some boundary condition that is slightly different across architectures?

If that's the case, I think a better fix would be to check that most of the values are greater than zero, rather than allow all values to be >= zero.

File: jwst/extract_1d/soss_extract/soss_extract.py
Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: Running RTs against a check on the model kwarg here: https://github.com/spacetelescope/RegressionTests/actions/runs/21756673452

File: jwst/extract_1d/tests/test_extract_1d_step.py (hunk 1 above)
Comment (tapastro) [reply jwst/extract_1d/tests/test_extract_1d_step.py]: See https://github.com/spacetelescope/jwst/pull/10235/changes/34161f32624d7640e8b6bd7df45a3fef3bc76f4b for an attempt to address this.

File: jwst/extract_1d/soss_extract/soss_extract.py
Comment (emolter) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: It looks like these regtests give only expected differences, except `test_extract_niriss_soss_96: jwst.extract_1d.tests.test_extract_1d_step` is still failing.  Remind me, was this run set off before you fixed that test?

The magnitude of the other regression test differences looks expected to me.  Although a relatively substantial fraction (10-40% or so) of the fluxes in the spectra are larger than the tolerance, most of the differences are small.  There are a few larger differences due to updates to bad pixel handling, which is why the "max relative difference" is large in some cases.  All of the white light differences are less than a tenth of a percent.

Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: Yes, sorry for the confusion - this RT run doesn't contain the unit test update. We can look at the latest CI for that.

Comment (melanieclarke) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: CI won't run that test because it's marked slow

Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: 🤦

**Review by emolter — APPROVED** [review]
Assuming your fix to that one slow-marked test actually works right, then this PR looks good to me.  I'll leave it up to you to decide if you want to rerun the regtest suite to ensure that test is passing.

File: jwst/extract_1d/soss_extract/soss_extract.py
Comment (melanieclarke) [reply jwst/extract_1d/soss_extract/soss_extract.py:1763]: You can use -k to run it by itself so you don't have to rerun the whole suite

**Comment by tapastro** [comment]
> Assuming your fix to that one slow-marked test actually works right, then this PR looks good to me. I'll leave it up to you to decide if you want to rerun the regtest suite to ensure that test is passing.

The first iteration failed - it appears as though the RT runner environment is producing some small amount of negative flux pixels. Moving the `np.all({FLUX} >= 0)` check to a 99% or more pixels >= 0 check passes now, but I don't know if we should be allowing negative flux pixels. Any opinion, @emolter ? We may need to catch these prior to saving the model.

**Comment by emolter** [comment]
@tapastro The fluxes in the spectra are reconstructed from a numerical retrieval.  Here, the mock input data has been set to unity everywhere.  I don't know how that should be expected to behave.  However, it appears that the "truth" files in `test_spec2_substrip96[x1dints]: jwst.regtest.test_niriss_soss` already contain negative values, which indicates that this was already a problem prior to Tim's PR and his PR just exposed it in this unit test.

My suggestion would be to leave the test as you modified it, merge this, and open a new issue to investigate why the fluxes can be negative under certain circumstances and how to fix it.
